internal/cache/cache.go

```diff
@@ -0,0 +1,111 @@
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+//go:generate mockgen -source=cache.go -package=cache -destination=mock_cache.go
+
+type Cache interface {
+	DeleteExpired()
+	Get(key string) (interface{}, bool)
+	Set(key string, value interface{})
+	StopCleanup()
+}
+
+type Item struct {
+	Object interface{}
+
+	Expiration int64
+}
+
+type cache struct {
+	mu         sync.RWMutex
+	items      map[string]Item
+	expiration time.Duration
+	stop       chan struct{}
+	wg         sync.WaitGroup
+}
+
+// New returns a new cache with the requested item expiration duration, which automatically
+// removes expired items at the cleanup interval.
+func New(expiration, cleanupInterval time.Duration) *cache {
+	c := &cache{
+		items:      make(map[string]Item),
+		expiration: expiration,
+		stop:       make(chan struct{}),
+	}
+
+	c.wg.Add(1)
+	go func(interval time.Duration) {
+		defer c.wg.Done()
+		c.cleanup(interval)
+	}(cleanupInterval)
+
+	return c
+}
+
+func (c *cache) cleanup(interval time.Duration) {
+	t := time.NewTicker(interval)
+
+	for {
+		select {
+		case <-t.C:
+			c.DeleteExpired()
+		case <-c.stop:
+			t.Stop()
+			return
+		}
+	}
+}
+
+// StopCleanup signals the cleanup goroutine to stop its ticker and return.
+func (c *cache) StopCleanup() {
+	close(c.stop)
+	c.wg.Wait()
+}
+
+// Set adds an item to the cache, replacing any existing item.
+func (c *cache) Set(key string, value interface{}) {
+	expiration := time.Now().Add(c.expiration).UnixNano()
+
+	c.mu.Lock()
+	c.items[key] = Item{
+		Object:     value,
+		Expiration: expiration,
+	}
+	c.mu.Unlock()
+}
+
+// Get fetches an item from the cache. It returns the item or nil and a bool indicating
+// whether the key was found.
+func (c *cache) Get(k string) (interface{}, bool) {
+	c.mu.RLock()
+	item, found := c.items[k]
+	if !found {
+		c.mu.RUnlock()
+		return nil, false
+	}
+	if item.Expiration > 0 {
+		if time.Now().UnixNano() > item.Expiration {
+			c.mu.RUnlock()
+			return nil, false
+		}
+	}
```

Comment:
Wondering why do we need this "double check". Shouldn't we treat all available items as valid until `DeleteExpired` is called?

Reply:
The reasoning behind this check is that I consider the expired items as expired, even if they are still in the cache until `DeleteExpired` is called. For example, item "a" has already expired but the `DeleteExpired` has not yet been called and a `Get("a")` arrives, which will not get the expired item back. WDYT?

Reply:
Yea, I got it from the code, just didn't expect it because we have the `DeleteExpired` method so I assume that everything is considered non-expired as long as it is in the cache but this is an implementation detail so if you prefer it that way I don't mind.